src/HttpClientFactory/Polly/src/DependencyInjection/PollyServiceCollectionExtensions.cs

```diff
@@ -105,4 +106,101 @@ public static IServiceCollection AddPolicyRegistry(this IServiceCollection servi
 
         return services;
     }
+
+    /// <summary>
+    /// Registers an empty <see cref="PolicyRegistry"/> in the service collection with service types
+    /// <see cref="IPolicyRegistry{String}"/>, <see cref="IReadOnlyPolicyRegistry{String}"/>, and
+    /// <see cref="IConcurrentPolicyRegistry{String}"/> if the service type hasn't already been registered
+    /// and returns the newly created registry.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+    /// <returns>The newly created <see cref="IPolicyRegistry{String}"/>.</returns>
+    public static IPolicyRegistry<string> TryAddPolicyRegistry(this IServiceCollection services)
+    {
+        if (services == null)
+        {
+            throw new ArgumentNullException(nameof(services));
+        }
+
+        // Create an empty registry, register and return it as an instance. This is the best way to get a
+        // single instance registered using all the interfaces.
+        var registry = new PolicyRegistry();
+
+        services.TryAddSingleton<IConcurrentPolicyRegistry<string>>(registry);
+        services.TryAddSingleton<IPolicyRegistry<string>>(registry);
+        services.TryAddSingleton<IReadOnlyPolicyRegistry<string>>(registry);
+
+        return registry;
+    }
+
+    /// <summary>
+    /// Registers the provided <see cref="IPolicyRegistry{String}"/> in the service collection with service types
+    /// <see cref="IPolicyRegistry{String}"/>, <see cref="IReadOnlyPolicyRegistry{String}"/>, and
+    /// <see cref="IConcurrentPolicyRegistry{String}"/> if the service type hasn't already been registered
+    /// and returns the provided registry.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+    /// <param name="registry">The <see cref="IPolicyRegistry{String}"/>.</param>
+    /// <returns>The provided <see cref="IPolicyRegistry{String}"/>.</returns>
+    public static IPolicyRegistry<string> TryAddPolicyRegistry(this IServiceCollection services, IPolicyRegistry<string> registry)
+    {
+        if (services == null)
+        {
+            throw new ArgumentNullException(nameof(services));
+        }
+
+        if (registry == null)
+        {
+            throw new ArgumentNullException(nameof(registry));
+        }
+
+        services.TryAddSingleton<IPolicyRegistry<string>>(registry);
```

Comment:
I have very little area knowledge, but should these additions be all-or-nothing?  Does it make sense to add the `IReadOnlyPolicyRegistry` if the `IPolicyRegistry` wasn't found (e.g. because there was already one registered)?

Reply:
I also don't have much knowledge of this area, but reading the existing code I saw that the policy registry must always be registered for all 3 interfaces (```IConcurrentPolicyRegistry```, ```IPolicyRegistry``` and ```IReadOnlyPolicyRegistry```).
Therefore, I would think there should never be a case where only one interface is registered. As you said, it's all or nothing.

In any case, I think the ```Add``` and ```TryAdd``` methods should register the same objects in the end. If we decide that the ```TryAdd```s should not be "all-or-nothing", we should also update the ```Add``` methods.